File: examples/connectors/custom_typescript_adapter/adapter.mjs
#!/usr/bin/env node
import readline from "node:readline"

const rl = readline.createInterface({ input: process.stdin, crlfDelay: Infinity })

function respond(id, ok, result, error = null) {
  process.stdout.write(`${JSON.stringify({ id, ok, result, error })}\n`)
}

for await (const line of rl) {
  if (!line.trim()) continue
  const request = JSON.parse(line)
  if (request.type === "shutdown") break
  if (request.type === "validate") {
    respond(request.id, true, { validated: true })
    continue
  }
  if (request.type !== "call") {
    respond(request.id, false, null, `unsupported request type ${request.type}`)
    continue
  }
  respond(request.id, true, {
    connector: request.connector,
    operation: request.operation,
    arguments: request.arguments ?? {},
    config: request.config ?? {},
    credential_id: request.credential?.id ?? null,
    has_secret: Boolean(request.credential?.secret),
  })
}

File: examples/connectors/custom_python_adapter/adapter.py
#!/usr/bin/env python3
import json
import sys


def respond(request_id, ok=True, result=None, error=None):
    print(json.dumps({"id": request_id, "ok": ok, "result": result, "error": error}), flush=True)


for line in sys.stdin:
    if not line.strip():
        continue
    request = json.loads(line)
    request_id = request.get("id", "")
    request_type = request.get("type")
    if request_type == "shutdown":
        break
    if request_type == "validate":
        respond(request_id, result={"validated": True})
        continue
    if request_type == "prepare":
        respond(request_id, result={
            "credential_targets": [],
            "prepared_config": request.get("config") or {},
        })
        continue
    if request_type != "call":
        respond(request_id, ok=False, error=f"unsupported request type {request_type}")
        continue
    credential = request.get("credential") or {}
    respond(request_id, result={
        "connector": request.get("connector"),
        "operation": request.get("operation"),
        "arguments": request.get("arguments") or {},
        "config": request.get("config") or {},
        "credential_id": credential.get("id"),
        "has_secret": bool(credential.get("secret")),
    })
